sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/InMemoryPartitionManager.cs

```diff
@@ -0,0 +1,155 @@
+            );
+        }
+
+        /// <summary>
+        ///   Tries to claim a list of specified ownership.
+        /// </summary>
+        ///
+        /// <param name="partitionOwnership">An enumerable containing all the ownership to claim.</param>
+        ///
+        /// <returns>An enumerable containing the successfully claimed ownership.</returns>
+        ///
+        public override Task<IEnumerable<PartitionOwnership>> ClaimOwnershipAsync(IEnumerable<PartitionOwnership> partitionOwnership)
+        {
+            return Task.Run(() =>
+                {
+                    List<PartitionOwnership> claimedOwnership = new List<PartitionOwnership>();
+
+                    foreach(var ownership in partitionOwnership)
+                    {
+                        var isClaimable = true;
+
+                        // In case the partition already has an owner, the ETags must match in order to claim it.
+
+                        if (Ownership.TryGetValue(ownership.PartitionId, out var currentOwnership))
+                        {
+                            isClaimable = ownership.ETag == currentOwnership.ETag;
+                        }
+
+                        // The following lock makes sure two different event processors won't try to claim ownership of a partition
+                        // simultaneously.  This approach prevents an ownership from being stolen just after being claimed.
+
+                        if (isClaimable)
+                        {
+                            lock (OwnershipClaimLock)
+                            {
+                                isClaimable = true;
+
+                                if (Ownership.TryGetValue(ownership.PartitionId, out currentOwnership))
+                                {
+                                    isClaimable = ownership.ETag == currentOwnership.ETag;
+                                }
+
+                                if (isClaimable)
+                                {
+                                    ownership.ETag = Guid.NewGuid().ToString();
+
+                                    Ownership[ownership.PartitionId] = ownership;
+                                    claimedOwnership.Add(ownership);
+
+                                    Log($"Ownership with partition id = '{ownership.PartitionId}' claimed.");
+                                }
+                                else
+                                {
+                                    Log($"Ownership with partition id = '{ownership.PartitionId}' is not claimable.");
+                                }
+                            }
+                        }
+                        else
+                        {
+                            Log($"Ownership with partition id = '{ownership.PartitionId}' is not claimable.");
+                        }
+                    }
+
+                    return claimedOwnership.AsEnumerable();
```

Comment:
There shouldn't be a need to `AsEnumerable` here;  List implements IEnumerable directly.